modules/tensor_mechanics/src/actions/TensorMechanicsAction.C

```diff
@@ -451,6 +469,140 @@ TensorMechanicsAction::actOutputGeneration()
   }
 }
 
+void
+TensorMechanicsAction::actEigenstrainNames()
+{
+  // Create containers for collecting blockIDs and eigenstrain names from materials
+  std::map<std::string, std::set<SubdomainID>> material_eigenstrain_map;
+  std::set<std::string> eigenstrain_set;
+
+  // Copy the current  vector of eigenstrain names
+  std::set<MaterialPropertyName> eigenstrain_names_copy(_eigenstrain_names.begin(),
+                                                        _eigenstrain_names.end());
+
+  std::set<MaterialPropertyName> verified_eigenstrain_names;
+
+  std::map<std::string, std::string> remove_add_map;
+  std::set<std::string> remove_reduced_set;
+
+  // Determine all the materials(eigenstrains) all ready created
+  auto materials = _problem->getMaterialWarehouse().getObjects();
+  for (auto & mat : materials)
+  {
+    const InputParameters & mat_params = mat->parameters();
+    auto & mat_name = mat->type();
+
+    // Check for eigenstrain names, only deal with those materials
+    if (mat_params.isParamValid("eigenstrain_name"))
+    {
+      std::shared_ptr<BlockRestrictable> blk = std::dynamic_pointer_cast<BlockRestrictable>(mat);
+      std::shared_ptr<MaterialData> mat_dat;
+      auto name = mat_params.get<std::string>("eigenstrain_name");
+
+      // Check for base_name prefix
+      if (mat_params.isParamValid("base_name"))
+      {
+        auto base_name = mat_params.get<std::string>("base_name");
+        name.insert(0, "_");
+        name.insert(0, base_name);
+      }
+
+      // Check block restrictions
+      if (blk)
+      {
+        const std::set<SubdomainID> & blocks =
+            blk->blockRestricted() ? blk->blockIDs() : blk->meshBlockIDs();
+        // Don't add duplicate
+        if (eigenstrain_set.insert(name).second)
+          material_eigenstrain_map.insert(std::make_pair(name, blocks));
+      }
+    }
+
+    // Account for reduced eigenstrains and CompositeEigenstrains
+    if (mat_params.isParamValid("input_eigenstrain_names"))
+    {
+      auto remove_list =
+          mat_params.get<std::vector<MaterialPropertyName>>("input_eigenstrain_names");
+      for (auto i : remove_list)
+        remove_reduced_set.insert(i);
+    }
+    // Account for CompositeEigenstrains
+    if (mat_name == "CompositeEigenstrain")
+    {
+      auto remove_list = mat_params.get<std::vector<MaterialPropertyName>>("tensors");
+      for (auto i : remove_list)
+      {
+        remove_reduced_set.insert(i);
+        if (material_eigenstrain_map.find(i) != material_eigenstrain_map.end())
+          material_eigenstrain_map.erase(i);
+      }
+    }
```

Comment:
This is neat, but needs _a lot_ of documentation (every class doc that is affected by this needs to link back to TMA with a not that it's input eigenstrains will be removed from automatic detection). Otherwise this has a huge potential for bugs to spring up in the future.

I think I'd also like a `mooseInfo` output of the final eigenstrain list if autodetection is enabled.